### PR TITLE
Create Optimism ovm2.get_contracts

### PR DESCRIPTION
**Pull request by MSilb7**
Creates tables to store and map contracts on Optimism. Check readmes for more information.

I've checked that:

* [x] the query produces the intended results
* [ ] the folder name matches the schema name
* [ ] the schema name exists in Dune
* [ ] views are prefixed with `view_`, functions with `fn_`.
* [ ] the filename matches the defined view, table or function and ends with .sql
* [ ] each file has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`

### CODE REVIEW

**Review by masquot — DISMISSED** [review]
You still have references to `dune_user_generated` in here. We can't deploy code that has these. Do you have a work-around?

**Comment by MSilb7** [comment]
> You still have references to `dune_user_generated` in here. We can't deploy code that has these. Do you have a work-around?

Got it, yeah I was going to do this in piecemeal, but I can add table creates for the `dune_user_generated` tables.

**Comment by MSilb7** [comment]
Pushed updates with a refactored version of the table + replaced dune user generated tables.

Still need to do some testing that the inserts work as expected before the next round.

**Comment by jeff-dude** [comment]
@MSilb7 based on recent commits, i assume you are ready for this one to be reviewed again & merged?

**Comment by MSilb7** [comment]
> @MSilb7 based on recent commits, i assume you are ready for this one to be reviewed again & merged?

Hi, I'm going to check one thing to see if I can use a recursive CTE vs hard coding 5 iterations of checking mappings. Once that's ready it should be good to review again!

**Comment by MSilb7** [comment]
Note: Updated to a new nft.tokens schema, pending https://github.com/duneanalytics/abstractions/pull/1000 getting merged.

(still not ready for review, just logging this)

**Comment by jeff-dude** [comment]
> Note: Updated to a new nft.tokens schema, pending #1000 getting merged.
> 
> (still not ready for review, just logging this)

fyi #1000 has been merged. i will continue to wait for your heads up to continue on this PR

**Comment by MSilb7** [comment]
> > Note: Updated to a new nft.tokens schema, pending #1000 getting merged.
> > (still not ready for review, just logging this)
> 
> fyi #1000 has been merged. i will continue to wait for your heads up to continue on this PR

This should be ready for first review now.

I'll save the recursive contract/factory mappings for a v2. Want to try to get. v1 in first since the traces table is starting to get large.

File: optimism2/ovm2/get_contracts/get_contracts.sql [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	created_time timestamptz,
+	contract_creator_if_factory bytea,
+	is_self_destruct boolean
+);
+
+CREATE INDEX IF NOT EXISTS ovm2_get_contracts_contract_address_idx ON ovm2.get_contracts (contract_address);
+CREATE INDEX IF NOT EXISTS ovm2_get_contracts_created_time_idx ON ovm2.get_contracts (created_time);
+CREATE INDEX IF NOT EXISTS ovm2_get_contracts_contract_project_idx ON ovm2.get_contracts (contract_project);
+CREATE INDEX IF NOT EXISTS ovm2_get_contracts_contract_address_contract_project_idx ON ovm2.get_contracts (contract_address,contract_project);
+CREATE INDEX IF NOT EXISTS ovm2_get_contracts_contract_address_contract_project_created_time_idx ON ovm2.get_contracts (contract_address,contract_project,created_time);
```
Comment (jeff-dude) [inline optimism2/ovm2/get_contracts/get_contracts.sql:18]: @MSilb7 index names have a max length of 63 characters in postgres. lines 18 & 19 are longer and get truncated, meaning they end up with the same name and error out. can you revisit naming standards and get all under 63 characters?

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: optimism2/ovm2/get_contracts/insert_contract_overrides.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+BEGIN;
+DELETE FROM ovm2.contract_overrides *;
+
+COPY ovm2.contract_overrides (contract_address,project_name,contract_name) FROM stdin;
+\\xc30141B657f4216252dc59Af2e7CdB9D8792e1B0	Socket	Socket Registry
+\.
+COMMIT;
+CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS contract_overrides_address_list_addr_uniq_idx ON ovm2.contract_overrides (contract_address);
+CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS contract_overrides_address_list_addr_proj_uniq_idx ON ovm2.contract_overrides (contract_address,project_name);
+CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS contract_overrides_address_list_addr_proj_contract_name_uniq_idx ON ovm2.contract_overrides (contract_address,project_name, contract_name);
```
Comment (jeff-dude) [inline optimism2/ovm2/get_contracts/insert_contract_overrides.sql:23]: same idea here, index name truncated to 63 characters. you may want to rename.

File: optimism2/ovm2/get_contracts/insert_updated_contract_info.sql [hunk 3]
```diff
@@ -0,0 +1,53 @@
+-- Scan for contracts that have a missing contract name or missing project name mapping and see if we have updated data to fill in.
+
+WITH contracts_to_update AS (
```
[the remark this answers is not in the document]
Comment (MSilb7) [reply optimism2/ovm2/get_contracts/insert_updated_contract_info.sql:3]: this is referenced in the insert below. It's intended to check if there are any contracts that we want to re-assign mappings for. Is it not functioning?

[the remark this answers is not in the document]
Comment (jeff-dude) [reply optimism2/ovm2/get_contracts/insert_updated_contract_info.sql:3]: using the WITH() CTE command, you're creating essentially a temp table for the current session, but it doesn't substantiate into the database. when you reference 'contracts_to_update' in a cron job, it won't recognize it. this query needs to either be a view or table (with corresponding function to load it). how would this object be used? if it's just manual lookup/reference, a view may make sense and be easier to maintain

[the remark this answers is not in the document]
Comment (MSilb7) [reply optimism2/ovm2/get_contracts/insert_updated_contract_info.sql:3]: Ok, this worked when I did non-cron inserts, but I'll make it a separate view.

It's essentially checking "of the contracts we have so far, are there any that we have a new mapping for & should update?"

File: optimism2/ovm2/get_contracts/get_contracts.sql (hunk 1 above)
Comment (MSilb7) [reply optimism2/ovm2/get_contracts/get_contracts.sql]: updated!

File: optimism2/ovm2/get_contracts/insert_contract_overrides.sql (hunk 2 above)
Comment (MSilb7) [reply optimism2/ovm2/get_contracts/insert_contract_overrides.sql]: updated

File: optimism2/ovm2/get_contracts/insert_updated_contract_info.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (MSilb7) [reply optimism2/ovm2/get_contracts/insert_updated_contract_info.sql]: updated

**Comment by MSilb7** [comment]
Updates should all be in!

File: optimism2/ovm2/get_contracts/insert_get_contracts.sql [hunk 4]
```diff
@@ -0,0 +1,302 @@
+CREATE OR REPLACE FUNCTION ovm2.insert_get_contracts(start_blocktime timestamptz, end_blocktime timestamptz, creator_list bytea[] = NULL::bytea[]) RETURNS integer
```
Comment (jeff-dude) [inline optimism2/ovm2/get_contracts/insert_get_contracts.sql:1]: i'm getting some syntax errors when attempting to create this function. looks like the CTE's are missing closing parentheses and/or too many commas. here is a head-start on the query in dune if you want to copy/paste into one and try to get it running as expected, then push changes here:
https://dune.com/queries/835559

Comment (MSilb7) [reply optimism2/ovm2/get_contracts/insert_get_contracts.sql:1]: oh weird yeah there were a few differences between this and what I was running on my end. Updating

**Comment by jeff-dude** [comment]
@MSilb7 i added one other commit for a minor fix. please refer to this query again for data 5/20/22 forward, let me know if the results look as expected, then we can get this function created and table loaded, close this PR out.
https://dune.com/queries/835559

**Comment by MSilb7** [comment]
> @MSilb7 i added one other commit for a minor fix. please refer to this query again for data 5/20/22 forward, let me know if the results look as expected, then we can get this function created and table loaded, close this PR out. https://dune.com/queries/835559

Hi, yes looks good!

**Review by jeff-dude — APPROVED** [review]
@MSilb7 i think we're all set here. as always, let me know if anything needs updated

**Pull request merged**